spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/GcpPubSubReactiveAutoConfiguration.java

```diff
@@ -49,30 +47,14 @@
 		matchIfMissing = true)
 public class GcpPubSubReactiveAutoConfiguration {
 
-	private Scheduler defaultPubSubReactiveScheduler;
-
 	@Bean
 	@ConditionalOnMissingBean
 	public PubSubReactiveFactory pubSubReactiveFactory(
 			PubSubSubscriberTemplate subscriberTemplate,
 			@Qualifier("pubSubReactiveScheduler") Optional<Scheduler> userProvidedScheduler) {
 
-		Scheduler scheduler = null;
-		if (userProvidedScheduler.isPresent()) {
-			scheduler = userProvidedScheduler.get();
-		}
-		else {
-			this.defaultPubSubReactiveScheduler = Schedulers.newElastic("pubSubReactiveScheduler");
-			scheduler = this.defaultPubSubReactiveScheduler;
-		}
+		Scheduler scheduler = userProvidedScheduler.orElseGet(() -> Schedulers.parallel());
```

Comment:
The reason why I opted for using `Schedulers.parallel()` is because it is optimised for non-blocking executions and this new code it is fully non-blocking, as you noted. 

I didn't use a named scheduler because IMHO in a non-blocking reactive environment, it is bad practice to use separate thread pools for separate tasks. Building reactive applications is about getting the most out of system resources. Separate thread pools would go against that. Having many more threads, meant for CPU work, then CPU cores is waste because they aren't doing anything for most of the time. They only cost memory and create more context-switching overhead.

At my company we have some applications that have thousands of threads because of the practice of using separate named thread pools for different tasks. Those applications are reaching their limits due to this and it is actually a problem.

I do understand your concern with regards to trouble shooting the bugs you describe. Without having named thread pools, it sometimes can be really difficult to figure out where the culprit is. But, TBH, in the reactive world dealing with those kind of issues is already especially difficult due to the nature of reactive programming. It would be nice if there was a "reactor native way" to dynamically enhance the name of thread, to add troubleshooting meta information...

These are just my 2 cents. I am not a specialist and maybe I am wrong about how important it is to do these kinds of optimisations. So, I will defer to your opinion on this. In the end the scheduler is configurable, so users can always optimise this however they want it to be.

Reply:
Never mind about what I said in the post above, because it isn't that relevant in this case. After look at it again, I noticed that the  `Schedulers.parallel()` scheduler is only used for doing the interval. It is not being used to handle the future, and the resulting Flux, because that is handled by the PubSub client thread that received the response. The `PubSubSubscriberTemplate.asyncPullExecutor` can be used to change the thread that handles that.

Reply:
@elefeint if you make it non-blocking then a custom scheduler here does not make sense at all and can be removed (`interval` will implicitly use `Schedulers.parallel()`) :)

Reply:
Excellent, thank you, this resolves my concern. It seems I have to unlearn the habit of creating a custom scheduler for every dedicated task.
The implication of `Flux.interval(Duration, Duration)` using `Schedulers.parallel()` by default is that we don't need a custom scheduler in `PubSubReactiveFactory` at all. On the other hand, it does not hurt anything to have extra flexibility in `pubSubReactiveScheduler` that the client application could control through overriding beans. Clearly I have not unlearned the habit of custom schedulers yet...

#2246 to track; this PR is large enough already.